pkg/types/clusterstate/clusterstate.go

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2020 Cortex Labs, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterstate
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/cortexlabs/cortex/pkg/lib/aws"
+	"github.com/cortexlabs/cortex/pkg/lib/errors"
+	"github.com/cortexlabs/cortex/pkg/lib/sets/strset"
+	s "github.com/cortexlabs/cortex/pkg/lib/strings"
+	"github.com/cortexlabs/cortex/pkg/lib/table"
+	"github.com/cortexlabs/cortex/pkg/types/clusterconfig"
+)
+
+const (
+	controlPlaneTemplate = "eksctl-%s-cluster"
+	operatorTemplate     = "eksctl-%s-nodegroup-ng-cortex-operator"
+	spotTemplate         = "eksctl-%s-nodegroup-ng-cortex-worker-spot"
+	onDemandTemplate     = "eksctl-%s-nodegroup-ng-cortex-worker-on-demand"
+)
+
+type ClusterState struct {
+	StatusMap    map[string]string // cloudformation stackname to cloudformation stackstatus
+	ControlPlane string
+	Nodegroups   []string
+	Status       Status
+}
+
+func any(statusMap map[string]string, statuses ...string) bool {
+	statusSet := strset.New(statuses...)
+	for _, stackStatus := range statusMap {
+		if statusSet.Has(stackStatus) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func all(statusMap map[string]string, statuses ...string) bool {
+	statusSet := strset.New(statuses...)
+	for _, stackStatus := range statusMap {
+		if !statusSet.Has(stackStatus) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (cs ClusterState) TableString() string {
+	var items table.KeyValuePairs
+	items.Add(cs.ControlPlane, cs.StatusMap[cs.ControlPlane])
+
+	for _, nodeGroupName := range cs.Nodegroups {
+		items.Add(nodeGroupName, cs.StatusMap[nodeGroupName])
+	}
+	return items.String()
+}
+
+func getStatus(statusMap map[string]string, controlPlane string) (Status, error) {
+	if statusMap[controlPlane] == string(StatusNotFound) {
+		return StatusNotFound, nil
+	}
+
+	if any(statusMap, cloudformation.StackStatusCreateFailed) {
+		return StatusCreateFailed, nil
+	}
+
+	if any(statusMap, cloudformation.StackStatusDeleteFailed) {
+		return StatusDeleteFailed, nil
+	}
+
+	if all(statusMap, cloudformation.StackStatusCreateComplete) {
```

Comment:
Should we also check that statusMap includes the info for the nodegroups? Is it possible if you run this when the cluster stack is complete and before the nodegoup stack creation is initiated, that it will say it's ready?